### PR TITLE
Bound declarative macro expansion growth

### LINKED ISSUE #972: [security][Medium] Declarative macro expansion bomb: total expanded size is unbounded (memory/CPU DoS)
## Severity: Medium

Declarative macro expansion is bounded only by the **number of passes** (`DEFAULT_MACRO_RECURSION_LIMIT = 64`), not by total expanded output size. A macro whose body expands to multiple self-invocations grows output geometrically per pass, exhausting memory/CPU well before the pass limit is reached.

## Locations

- `stage1/crates/axiomc/src/syntax.rs:59` — `DEFAULT_MACRO_RECURSION_LIMIT = 64`
- `:1117` — `for pass in 0..options.macro_recursion_limit` (the only bound is pass count)
- `:1544-1560` — `expand_macro_invocations_once`
- `:1640-1656` — statement-context multi-line expansion

There is no cap on cumulative expanded line/byte count or total invocation count.

## Exploit

A crafted `.ax` file defining a macro that expands into two statement-context calls of itself: each pass roughly doubles the output, so after a few dozen passes the source has grown by many orders of magnitude. Building/checking it exhausts memory and CPU.

## Suggested fix

- Track and cap cumulative expanded **byte/line count** (and/or total invocation count) across all passes, returning a diagnostic when the budget is exceeded — not just the per-pass-depth limit.

---
_Filed as part of a comprehensive security review. No code changes were made._

### PR DESCRIPTION
**Pull request by jmcte**
## Summary

- Add default declarative macro expansion budgets for expanded source bytes and cumulative invocation count.
- Validate macro expansion options and report parser diagnostics when a budget is invalid or exceeded.
- Add a regression test for a self-amplifying macro that grows output before the recursion-depth limit is reached.

## Governing Issue

Closes #972

Refs #973

## Validation

- [x] Relevant local checks passed
- [x] Required PR checks are expected to satisfy `CI Gate`
- [x] Skipped checks are explained below

Commands run:

- `cargo test --manifest-path stage1/Cargo.toml -p axiomc parser_bounds_total_declarative_macro_expansion_size`
- `cargo test --manifest-path stage1/Cargo.toml -p axiomc parser_bounds_recursive_declarative_macro_expansion`
- `cargo test --manifest-path stage1/Cargo.toml -p axiomc parser_honors_macro_recursion_limit_option`
- `cargo test --manifest-path stage1/Cargo.toml -p axiomc macro`
- `git diff --check -- stage1/crates/axiomc/src/syntax.rs stage1/crates/axiomc/src/lib.rs stage1/crates/axiomc/src/project.rs`

Skipped checks:

- Full workspace test suite was not run; this patch is scoped to declarative macro expansion and the macro-filtered `axiomc` tests passed.

## Bootstrap Governance

- [x] Changes are scoped to the linked issue
- [x] Contributor or PR guidance changes are reflected in `CONTRIBUTING.md`, `.github/PULL_REQUEST_TEMPLATE.md`, and `docs/bootstrap/onboarding.md` when applicable
- [ ] Auto-merge is enabled, or GitHub plan-limit evidence is recorded and the fallback merge-readiness policy applies
- [x] No real secrets, runtime auth, or machine-local env files are committed

Auto-merge note: not enabled from this branch; fallback merge-readiness applies once checks pass and approval/conversation requirements are satisfied.

## Semantic Layer Checklist

- [x] Semantic node types added/changed are listed, or this PR does not touch semantic-layer behavior
- [x] Schemas added/changed are listed, or this PR does not touch schemas
- [x] Evidence added/changed is listed, or this PR does not touch evidence
- [x] Rust capture check is satisfied, or this PR is backend-only and marks backend-specific behavior
- [x] Agent-facing inspection impact is described, or there is no inspection impact

Semantic layer impact:

- Node types added/changed: none.
- Schemas added/changed: none.
- Evidence added/changed: none.
- Rust capture risk: reduced in the current Rust parser by bounding declarative macro expansion growth.
- Agent-facing inspection impact: parse/check/graph flows fail fast with a diagnostic instead of allowing macro expansion to consume unbounded memory or CPU.

## Flow Contract

- Owner lane: compiler/runtime
- Repair owner: Codex
- Autonomy class: autonomous issue repair
- Risk class: Medium security hardening

## Flow Merge Readiness

- [x] Every blocker has a next actor and next action
- [x] No active blocking requested changes remain
- [ ] Non-author approval is present when required
- [ ] Auto-merge is appropriate when gates pass

## Merge Automation

- [ ] Auto-merge is enabled, or the reason it is unavailable or unsafe is noted below

Auto-merge is not enabled because maintainer approval and branch-protection gates must complete first.

## Notes

- The default expanded-source budget is 16 MiB per module and the default cumulative invocation budget is 8192 expansions.
- Existing macro recursion-depth behavior remains intact.

### CODE REVIEW

File: stage1/crates/axiomc/src/syntax.rs [hunk 1]
```diff
@@ -1123,7 +1146,25 @@ fn expand_declarative_macros(
             invocation_chain.push(format!("{}!", expansion.macro_name));
         }
         expansions.extend(line_expansions);
+        if expansions.len() > options.macro_expansion_invocation_limit {
```
[the remark this answers is not in the document]
Comment (jmcte) [reply stage1/crates/axiomc/src/syntax.rs:1149]: @codex please address

**Review by athena-omt — DISMISSED** [review]
The new macro-expansion budget logic is close, but there is one correctness bug that will reject valid code at the exact budget boundary.

- [stage1/crates/axiomc/src/syntax.rs](file:///openclaw-data/src/axiom/stage1/crates/axiomc/src/syntax.rs#L1213-L1221): `validate_macro_expansion_byte_budget()` always adds `1` for a newline after every line, including the last one. But `expanded_lines_to_source()` joins lines with `"\n"` and does not append a trailing newline, so the validator overcounts by one byte. Any expansion whose rendered size is exactly equal to the configured limit will be reported as over budget. That makes the default 16 MiB cap and any custom limit slightly too strict at the boundary.

Please fix the accounting to match the actual rendered source size, and add a regression test that lands exactly on the limit.

Compact summary: one boundary-condition bug in the new byte-budget enforcement; otherwise the shape of the change looks sensible.

**Comment by jmcte** [comment]
Addressed Athena's boundary feedback in 4606a50d: the byte-budget validator now counts newlines only between expanded lines, matching expanded_lines_to_source(), and there is a regression test for an expansion that lands exactly on the byte limit. Local validation: cargo test --manifest-path stage1/Cargo.toml -p axiomc parser_allows_declarative_macro_expansion_at_exact_byte_limit; cargo test --manifest-path stage1/Cargo.toml -p axiomc macro; git diff --check -- stage1/crates/axiomc/src/syntax.rs stage1/crates/axiomc/src/lib.rs.

**Review by pheidon — APPROVED** [review]
Reviewed the current head commit. The macro-expansion byte-budget boundary now matches the rendered source size, and the regression coverage looks right. LGTM.
